### PR TITLE
Replace ugorji/codec with json-iterator/go

### PR DESCRIPTION
**Pull request by dims**
Kubernetes removed the usage for ugorji in the following PR:
[Use json-iterator instead of ugorji for JSON. #48287](https://github.com/kubernetes/kubernetes/pull/48287)

However since etcd still uses this library, we end up increasing the size of the binaries in kubernetes by a lot. See:
https://github.com/kubernetes/kubernetes/issues/76506

Can we please switch over to json-iterator/go as well to eliminate the unnecessary code/size?

Change-Id: Ia8b267e9ada49b22a1cd3bd7c05450e32582dc91

Please read https://github.com/etcd-io/etcd/blob/master/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by dims** [comment]
@gyuho @spzala any idea how i can get the `amd64-e2e` of the `semaphoreci` to pass?

**Comment by spzala** [comment]
@dims I don't see any failure related to your changes looking at quickly. Lately, CI is flakey. I will be reviewing the changes too. Thanks!

**Comment by liggitt** [comment]
github.com/ugorji/go is still listed in go.sum and is vendored... did you run `go mod tidy && go mod vendor`?

**Comment by dims** [comment]
@liggitt unfortunately the `scripts/updatedep.sh` does not seem to work well, it just nukes everything in `go.mod` and `go.sum`. So i am hesitating to add a commit with the cleanup

**Comment by dims** [comment]
@liggitt cleaned up as well! dropped 45k+ LOC

**Comment by gyuho** [comment]
@dims Could you update the commit title to `Expected commit title format '<package>{", "<package>}: <description>'`?

File: client/json.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		iter.ReportError("DecodeNumber", err.Error())
+	default:
+		*(*interface{})(ptr) = iter.Read()
+	}
+}
+
+// CaseSensitiveJsonIterator returns a jsoniterator API that's configured to be
+// case-sensitive when unmarshalling, and otherwise compatible with
+// the encoding/json standard library.
+func CaseSensitiveJsonIterator() jsoniter.API {
```
Comment (liggitt) [inline client/json.go]: suggest not exporting this if you don't have to

File: client/keys.go [hunk 2]
```diff
@@ -658,7 +654,7 @@ func unmarshalHTTPResponse(code int, header http.Header, body []byte) (res *Resp
 
 func unmarshalSuccessfulKeysResponse(header http.Header, body []byte) (*Response, error) {
 	var res Response
-	err := codec.NewDecoderBytes(body, new(codec.JsonHandle)).Decode(&res)
+	err := CaseSensitiveJsonIterator().Unmarshal(body, &res)
```
Comment (liggitt) [inline client/keys.go]: can we reuse a private iterator instance?

File: client/json.go (hunk 1 above)
Comment (dims) [reply client/json.go]: Ack doing that now

File: client/keys.go (hunk 2 above)
Comment (dims) [reply client/keys.go]: ack doing that now.

**Comment by dims** [comment]
@gyuho ack will fix the commit message shortly

**Comment by dims** [comment]
@gyuho @spzala would you consider a backport to release-3.3 branch? (looks like k/k is using v3.3.10+ currently)

**Comment by gyuho** [comment]
@dims Sure, if it doesn't break compatibility. Would anything change in `json-iterator/go`?

**Comment by spzala** [comment]
@dims aha..all green checks, looks good :-)

**Comment by dims** [comment]
@gyuho exact same code in this PR applies cleanly to 3.3-release (except for the vanity url). So we should be good

**Pull request merged**